### PR TITLE
proposal for console to be able to close without closing entire Shoes

### PR DESCRIPTION
**Pull request by passenger94**
could you please have a look around this especially :
https://github.com/passenger94/shoes3/blob/video_vlc3/shoes/console/tesi.c#L550

### CODE REVIEW

**Comment by ccoupe** [comment]
What is the world are you using that terminal code for ? Ah -- puts output from tests because it capture stdout and stderr?  Clever devil.

**Comment by BackOrder** [comment]
What would be the point of this feature? It sounds like plain Ruby to me.

**Comment by ccoupe** [comment]
@BackOrder , I'm guessing your comment belongs on a different thread #212 perhaps.

**Comment by BackOrder** [comment]
My comment is pertaining to the title of this thread. As far as #212 goes, it might make sense but then I would suggest to have a special switch (i.e. shoes.exe --ruby) to tell shoes it's running in a Ruby console mode.

**Comment by passenger94** [comment]
@BackOrder "console" in the title is probably misleading, i'm talking about the nice terminal @ccoupe crafted for our poor souls to feel comfortable in the general tormented flow of the universe and anecdotally, to capture output of test-units while shoes is already running

**Comment by ccoupe** [comment]
OK. We've had  'cshoes --console script.rb ' or 'cshoes.exe -w script.rb' or platform variations of  that going back to 3.2.24. Is that a different console than the one you're thinking of. It has to be invoked with an explicit `Shoes::show_console' or from the command line and behaves much differently than  the Shoes console you use for  info, warn, debug etc. The console  @passenger94  fixed  is damn near independent  of Shoes and Ruby.

**Comment by BackOrder** [comment]
@passenger94  This could work. It's possible to write your own test runner interface. This is how (G)UI Test Runners are written. Ruby makes it fairly easy.

**Comment by passenger94** [comment]
@BackOrder look at #210, i've made this working and added some ideas to develop ...

**Comment by ccoupe** [comment]
perhaps samples/expert-console.rb will be informative.

**Comment by passenger94** [comment]
@ccoupe i'm trying to capture escape sequence from unit test to show_console in order to colorize output
the colorize part is working( i can put some colorized text in console), but from my test it seems that no escape character are going into the console ... they are lost somewhere into the way from test units and the console .....
in **console_haveChar** 
```c
case '\e': //  same as '\x1B',  but that's what test units sends
      gtk_text_buffer_get_end_iter(buffer, &iter_e);
// here i'm expecting a red "S" in console if i capture a CSI, but nothing
      gtk_text_buffer_insert_with_tags(buffer, &iter_e, "S", -1, redTag, NULL);
      tobj->partialSequence = 1;
      break;
     ...........
``` 
Could you shed some lights on this ? 
(i think i'm ok with the escape sequence thingy and think i can find my path in tesi code (ancient and yours),  but the tty/Master/Slave is cryptic to me)

**Comment by ccoupe** [comment]
The escape handling code in tesi was broken in several ways.  Beware of what happens with Windows - it doesn't use that code. OSX doesn't use that code. Only Windows will respond to escape sequences and it probably uses different ones than your used to using.

**Comment by passenger94** [comment]
OoooKay ! damn, if one think life is boring, just explore cross platform coding ...

**Comment by passenger94** [comment]
i went very minimal and ended up using only your **console_haveChar** which is present in both gtk and cocoa ...
Because i had a hard time finding how to catch the escape sequence from test unit, the very only thing i'm trying to do right now is just trying to cath the CSI, in which i failed, nothing, there seems to be no CSI going to the console, so i'm wondering if it is transmitted from the tty/Master/Slave ceremony, test unit uses the console runner, which by default writes to STDOUT and the shown results proves that the connection is working ....  ???
I don't know if it is doable on all platform, but i need very few information, just the sequence that test units sends (assuming cocoa textview equivalent is able to colorize text)

**Comment by ccoupe** [comment]
What you call CSI is a mult-character string starting with ESC, usuallly followed by '[' but not always, followed by another char and then some other characters that depend on earliers and then a table/hash thing to find the color in Gtk terms.  gtk-terminal.c line 69 where you start.

The unused tesi code doesn't handle color setting sequences but it  shows how to use an ad-hoc state machine to handle escape sequences.  Try your tests on Windows before you get  deep in coding.  That one can't be coded around because it is a MSFT api.  - Windows doesn't have pty like unix does and Shoes can create one - I looked into - big hack to Windows system.

Setting color on Cocoa is not too hard if I remember. Keeping the state machine accurate is a big deal. Providing a terminal descriptor (which is probably what rtest is trying to use) -- Very big problem. I looked into it - I've written terminal emulators before and parsed those escape characters.  Tesi was written for a different purpose than our. It was limited then, it was buggy then. 

it would be easier to convince rtest to not colorize.

**Comment by passenger94** [comment]
if i do this in a simple Shoes app, with console open, it works :
`$stdout.write "not colorized this yes -->\e32Km"`

so console is not the problem (apart cleaning ancient tesi code, for test unit it's not needed)

we can totally ignore the colorization from tests, not a problem, but psychological impact of green nowadays ....

**Comment by ccoupe** [comment]
\e[32m (and its red equivalent  and the return to default),  should be a fairly small state machine to add to gtk-terminal.c:69  Detect escape - set state=1. After line 63 check for state==1, and if char is '[', set state = 2. If state==3 and char is '3' .....  Remember to reset state = 0 when you have the last char, then you need to call gtk with what it expect for color. Simples. It's the cursor moves & that cause the complexity.

**Comment by passenger94** [comment]
:-D, ok, let me rephrase, that part works !, it's just that console receives no escape sequence from test units (it's ok when test units uses the OS terminal) and i wonder why ! i'm suspecting the sequence is trapped in the tty/master/slave wiring, but that thing is way over me.

**Comment by ccoupe** [comment]
Most likely , it's the absence of a terminal definition attached to the gtk-terminal or rtest know's it talking to a pty instead of a tty. Or it notices that ENV[TERM] is 'vt102' which doesn't support colors or it sent an escape sequence to the terminal to identify what's there and got nothing back.  Terminal identification  is very strange in unix if you've never used it. You could try setting ENV[TERM]='xterm' or not set it at all in gtk-terminal.c but my guess is that won't help.

**Comment by passenger94** [comment]
Boom ! right in the middle of the center ! Thanks ! 'xterm' was the secret pass, let's see what Windows think about it now (not optimist)

**Comment by ccoupe** [comment]
Windows is weird. (Shoes doesn't use the gtk-terminal.c/tesi.c on Windows) I think the code is at the end of gtk.c. It launches a new dos command window but only if you didn't start Shoes from the command line - then it uses the exist terminal.  I wish OSX and Linux could do that too.

make sure samples/expert-console.rb works. 

Windows has no concept of tty/pty access or remote terminals (hence no pty). It's just bad by anything we know.  In an msys window
```
Cecil@bronco7 /e/exe-shoes
$ echo $TERM
cygwin
```
In a dos command window. there is no ENV[TERM] or anything like it. Msys shell (cygwin) sucks just as bad.
